lms/services/roster.py

```diff
@@ -373,21 +504,105 @@ def _get_roster_roles(self, roster) -> list[LTIRole]:
         roles = {role for member in roster for role in member["roles"]}
         return self._lti_role_service.get_roles(list(roles))
 
-    def _get_lti_registration(self, lms_course) -> LTIRegistration:
-        lti_registration = self._db.scalars(
-            select(LTIRegistration)
-            .join(ApplicationInstance)
-            .where(LMSCourseApplicationInstance.lms_course_id == lms_course.id)
+    def _get_application_instance(self, lms_course) -> ApplicationInstance:
+        return self._db.scalars(
+            select(ApplicationInstance)
             .join(LMSCourseApplicationInstance)
-            .order_by(LTIRegistration.updated.desc())
+            .where(LMSCourseApplicationInstance.lms_course_id == lms_course.id)
+            .order_by(LMSCourseApplicationInstance.updated.desc())
         ).first()
-        assert lti_registration, "No LTI registration found for LMSCourse."
-        return lti_registration
+
+    def _get_lti_registration(self, lms_course) -> LTIRegistration:
+        ai = self._get_application_instance(lms_course)
+        assert ai.lti_registration, "No LTI registration found for LMSCourse."
+        return ai.lti_registration
+
+    def _get_course_instructor(self, lms_course: LMSCourse) -> LMSUser | None:
+        return self._db.scalars(
+            select(LMSUser)
+            .join(LMSCourseMembership)
+            .join(LTIRole)
+            .where(
+                LMSCourseMembership.lms_course_id == lms_course.id,
+                LTIRole.type == RoleType.INSTRUCTOR,
+                LTIRole.scope == RoleScope.COURSE,
+            )
+            .order_by(LMSCourseMembership.updated.desc())
+        ).first()
+
+    def _get_canvas_sections(
+        self,
+        canvas_api_client: CanvasAPIClient,
+        oauth2_token_service: OAuth2TokenService,
+        lms_course: LMSCourse,
+        with_refresh_token=False,
```

Comment:
Perhaps you can set this as `True` by default. Then you only have to change the value the second time, when internally retrying.